File: src/iOS/OSMModels/OsmMember.m
//
//  OsmMember.m
//  Go Map!!
//
//  Created by Wolfgang Timme on 1/18/20.
//  Copyright © 2020 Bryce. All rights reserved.
//

#import "OsmMember.h"

@implementation OsmMember

-(NSString *)description
{
    return [NSString stringWithFormat:@"%@ role=%@; type=%@;ref=%@;", [super description], _role, _type, _ref ];
}
-(id)initWithType:(NSString *)type ref:(NSNumber *)ref role:(NSString *)role
{
    self = [super init];
    if ( self ) {
        _type = type;
        _ref = ref;
        _role = role;
    }
    return self;
}
-(id)initWithRef:(OsmBaseObject *)ref role:(NSString *)role
{
    self = [super init];
    if ( self ) {
        _ref = ref;
        _role = role;
        if ( ref.isNode )
            _type = @"node";
        else if ( ref.isWay )
            _type = @"way";
        else if ( ref.isRelation )
            _type = @"relation";
        else {
            _type = nil;
        }
    }
    return self;
}

-(void)resolveRefToObject:(OsmBaseObject *)object
{
    assert( [_ref isKindOfClass:[NSNumber class]] || [_ref isKindOfClass:[OsmBaseObject class]] );
    assert( [object isKindOfClass:[NSNumber class]] || (object.isNode && self.isNode) || (object.isWay && self.isWay) || (object.isRelation && self.isRelation) );
    _ref = object;
}


-(BOOL)isNode
{
    return [_type isEqualToString:@"node"];
}
-(BOOL)isWay
{
    return [_type isEqualToString:@"way"];
}
-(BOOL)isRelation
{
    return [_type isEqualToString:@"relation"];
}

-(void)encodeWithCoder:(NSCoder *)coder
{
    OsmBaseObject * o = _ref;
    NSNumber * ref = [_ref isKindOfClass:[OsmBaseObject class]] ? o.ident : _ref;
    [coder encodeObject:_type    forKey:@"type"];
    [coder encodeObject:ref        forKey:@"ref"];
    [coder encodeObject:_role    forKey:@"role"];
}
-(id)initWithCoder:(NSCoder *)coder
{
    self = [super init];
    if ( self ) {
        _type    = [coder decodeObjectForKey:@"type"];
        _ref    = [coder decodeObjectForKey:@"ref"];
        _role    = [coder decodeObjectForKey:@"role"];
    }
    return self;
}

@end

File: src/Shared/OSMModels/OsmMember.swift
//
//  OsmMember.swift
//  Go Map!!
//
//  Created by Wolfgang Timme on 1/18/20.
//  Copyright © 2020 Bryce Cogswell. All rights reserved.
//

final class OsmMember: NSObject, NSSecureCoding {
	static let supportsSecureCoding: Bool = true

	let ref: OsmIdentifier
	private(set) var type: OSM_TYPE // way, node, or relation
	private(set) var obj: OsmBaseObject?
	private(set) var role: String?

	override var description: String {
		return "\(super.description) role=\(role ?? ""); type=\(type); ref=\(ref);"
	}

	init(type: OSM_TYPE, ref: OsmIdentifier, role: String?) {
		self.type = type
		self.ref = ref
		obj = nil
		self.role = role
		super.init()
	}

	init(obj: OsmBaseObject, role: String?) {
		self.obj = obj
		ref = obj.ident
		self.role = role
		if obj.isNode() != nil {
			type = .NODE
		} else if obj.isWay() != nil {
			type = .WAY
		} else if obj.isRelation() != nil {
			type = .RELATION
		} else {
			fatalError()
		}
		super.init()
	}

	func deresolveRef() {
		obj = nil
	}

	func resolveRef(to object: OsmBaseObject) {
		precondition(ref == object.ident)
		obj = object
	}

	func isNode() -> Bool {
		return type == .NODE
	}

	func isWay() -> Bool {
		return type == .WAY
	}

	func isRelation() -> Bool {
		return type == .RELATION
	}

	func encode(with coder: NSCoder) {
		coder.encode(type.string, forKey: "type")
		coder.encode(NSNumber(value: ref), forKey: "ref")
		coder.encode(role, forKey: "role")
	}

	required init?(coder: NSCoder) {
		guard let type2 = coder.decodeObject(forKey: "type") as? String,
		      let type = try? OSM_TYPE(string: type2)
		else {
			return nil
		}
		self.type = type
		guard let ref2 = coder.decodeObject(forKey: "ref")
		else { fatalError("OsmMember ref is nil") }
		if let ref2 = ref2 as? NSNumber {
			// normal path
			ref = ref2.int64Value
		} else if let ref2 = ref2 as? OsmBaseObject {
			// shouldn't happen but seems to when upgrading from old obj-c versions
			ref = ref2.ident
		} else {
			// shouldn't happen
			fatalError("OsmMember ref is not NSNumber: \(Swift.type(of: ref2))")
		}
		role = coder.decodeObject(forKey: "role") as? String
		obj = nil
		super.init()
	}
}
